op_softmax.go

```diff
@@ -0,0 +1,272 @@
+func (op *softmaxOp) DoDiff(ctx ExecutionContext, inputs Nodes, output *Node) error {
+	if len(inputs) != 1 {
+		return fmt.Errorf("SoftmaxOp.DoDiff needs 1 arguments")
+	}
+
+	odv := output.boundTo.(*dualValue)
+	odvd := odv.Value.(tensor.Tensor)
+	diffOp := newSoftmaxOpDiff()
+
+	result, err := diffOp.Do()
+	if err != nil {
+		return err
+	}
+
+	sum, err := odvd.(*tensor.Dense).Add(result.(*tensor.Dense), tensor.UseUnsafe())
+	if err != nil {
+		return err
+	}
+
+	odv.d = sum
+
+	return nil
+}
+
+// SymDiff applies the diff op. Implementation for SDOp interface.
+func (op *softmaxOp) SymDiff(inputs Nodes, output, grad *Node) (Nodes, error) {
+	err := checkArity(op, len(inputs))
+	if err != nil {
+		return nil, err
+	}
+
+	diffOp := newSoftmaxOpDiff()
+	nodes := make(Nodes, 1)
+
+	nodes[0], err = ApplyOp(diffOp, output)
+
+	return nodes, err
+}
+
+// DiffWRT is an implementation for the SDOp interface
+func (op *softmaxOp) DiffWRT(inputs int) []bool {
+	if inputs != 1 {
+		panic(fmt.Sprintf("softmax operator only supports one input, got %d instead", inputs))
+	}
+
+	return []bool{true}
+}
+
+type softmaxDiffOp struct {
+}
+
+func newSoftmaxOpDiff() *softmaxDiffOp {
+	return &softmaxDiffOp{}
+}
+
+func (op *softmaxDiffOp) Arity() int {
+	return 1
+}
+
+func (op *softmaxDiffOp) ReturnsPtr() bool { return false }
+
+func (op *softmaxDiffOp) CallsExtern() bool { return false }
+
+func (op *softmaxDiffOp) WriteHash(h hash.Hash) {
+	fmt.Fprintf(h, "SoftmaxDiff{}()")
+}
+
+func (op *softmaxDiffOp) Hashcode() uint32 { return simpleHash(op) }
+
+func (op *softmaxDiffOp) String() string {
+	return fmt.Sprintf("SoftmaxDiff{}()")
+}
+
+func (op *softmaxDiffOp) InferShape(inputs ...DimSizer) (tensor.Shape, error) {
+	s := inputs[0].(tensor.Shape).Clone()
+
+	return s, nil
+}
+
+func (op *softmaxDiffOp) Type() hm.Type {
+	aType := hm.TypeVariable('a')
+
+	ta := newTensorType(1, aType)
+
+	return hm.NewFnType(ta, ta) // f(float64) float64
+}
+
+func (op *softmaxDiffOp) OverwritesInput() int { return -1 }
+
+func (op *softmaxDiffOp) checkInput(inputs ...Value) (tensor.Tensor, error) {
+	if err := checkArity(op, len(inputs)); err != nil {
+		return nil, err
+	}
+
+	var (
+		in tensor.Tensor
+
+		ok bool
+	)
+
+	switch t := inputs[0].(type) {
+	case *dualValue:
+		if in, ok = t.Value.(tensor.Tensor); !ok {
+			return nil, errors.Errorf("input should be a tensor, got %T", inputs[0])
+		}
+	case tensor.Tensor:
+		in = t
+	default:
+		return nil, errors.Errorf("input type is not supported, got %T", inputs[0])
+	}
+
+	return in, nil
+}
+
+func (op *softmaxDiffOp) Do(inputs ...Value) (Value, error) {
+	inputTensor, err := op.checkInput(inputs...)
+	if err != nil {
+		return nil, fmt.Errorf("Can't check SoftmaxDiff input: %w", err)
+	}
+
+	diag := tensor.New(tensor.AsDenseDiag(inputTensor))
+
+	sm := inputTensor.Clone().(tensor.Tensor)
+	sm.Reshape(inputTensor.Shape().TotalSize(), 1)
+
+	smT := sm.Clone().(tensor.Tensor)
+	smT.Transpose()
```

Comment:
you can use `smT.T()`, which does no work, rather than `Transpose` which does a lot of work